src/ui/handler/touch_zoom_rotate.js

```diff
@@ -1,293 +1,306 @@
+    for (let i = 0; i < e.targetTouches.length; i++) {
+        if (e.targetTouches[i].identifier === identifier) return points[i];
     }
+}
+
+/* ZOOM */
 
-    onStart(e: TouchEvent) {
-        if (!this.isEnabled()) return;
-        if (e.touches.length !== 2) return;
+const ZOOM_THRESHOLD = 0.1;
 
-        const p0 = DOM.mousePos(this._el, e.touches[0]),
-            p1 = DOM.mousePos(this._el, e.touches[1]),
-            center = p0.add(p1).div(2);
+function getZoomDelta(distance, lastDistance) {
+    return Math.log(distance / lastDistance) / Math.LN2;
+}
+
+export class TouchZoomHandler extends TwoTouchHandler {
 
-        this._startVec = p0.sub(p1);
-        this._startAround = this._map.transform.pointLocation(center);
-        this._gestureIntent = undefined;
-        this._inertia = [];
+    _distance: number;
+    _startDistance: number;
 
-        DOM.addEventListener(window.document, 'touchmove', this._onMove, {passive: false});
-        DOM.addEventListener(window.document, 'touchend', this._onEnd);
+    reset() {
+        super.reset();
+        delete this._distance;
+        delete this._startDistance;
     }
 
-    _getTouchEventData(e: TouchEvent) {
-        const p0 = DOM.mousePos(this._el, e.touches[0]),
-            p1 = DOM.mousePos(this._el, e.touches[1]);
+    _start(points: [Point, Point]) {
+        this._startDistance = this._distance = points[0].dist(points[1]);
+    }
 
-        const vec = p0.sub(p1);
+    _move(points: [Point, Point], pinchAround: Point) {
+        const lastDistance = this._distance;
+        this._distance = points[0].dist(points[1]);
+        if (!this._active && Math.abs(getZoomDelta(this._distance, this._startDistance)) < ZOOM_THRESHOLD) return;
+        this._active = true;
         return {
-            vec,
-            center: p0.add(p1).div(2),
-            scale: vec.mag() / this._startVec.mag(),
-            bearing: this._rotationDisabled ? 0 : vec.angleWith(this._startVec) * 180 / Math.PI
+            zoomDelta: getZoomDelta(this._distance, lastDistance),
+            pinchAround
         };
     }
+}
 
-    _onMove(e: TouchEvent) {
-        if (e.touches.length !== 2) return;
+/* ROTATE */
 
-        const {vec, scale, bearing} = this._getTouchEventData(e);
+const ROTATION_THRESHOLD = 25; // pixels along circumference of touch circle
 
-        // Determine 'intent' by whichever threshold is surpassed first,
-        // then keep that state for the duration of this gesture.
-        if (!this._gestureIntent) {
-            // when rotation is disabled, any scale change triggers the zoom gesture to start
-            const scalingSignificantly = (this._rotationDisabled && scale !== 1) || (Math.abs(1 - scale) > significantScaleThreshold),
-                rotatingSignificantly = (Math.abs(bearing) > significantRotateThreshold);
+function getBearingDelta(a, b) {
+    return a.angleWith(b) * 180 / Math.PI;
```

Comment:
Random note as a I look through the code again, it would be worth having these as defined as constants somewhere so we would write this instead `a.angleWith(b) * Util.DEG2RAD`, we're doing this in a few spots: https://github.com/mapbox/mapbox-gl-js/search?q=%22180+%2F+Math.PI%22&unscoped_q=%22180+%2F+Math.PI%22.

Not to address in this PR though.

Reply:
Yep, agreed!